inbm/dispatcher-agent/dispatcher/sota/sota.py

```diff
@@ -239,12 +250,16 @@ def execute(self, proceed_without_rollback: bool, skip_sleeps: bool = False) ->
         setup_helper = self.factory.create_setup_helper()
         if self.sota_cmd == 'rollback':
             self.snap_num = setup_helper.get_snapper_snapshot_number()
+            self._update_logger.detail_status = ROLLBACK
         snapshot = self.factory.create_snapshotter(
             self.sota_cmd, self.snap_num, self.proceed_without_rollback, self._is_reboot_device())
         rebooter = self.factory.create_rebooter()
 
         if self.sota_state == 'diagnostic_system_unhealthy':
             self._update_logger.update_log(FAIL)
+            self._update_logger.detail_status = ROLLBACK
+            self._update_logger.error = "Critical service failure."
+            self.save_granular_log(check_package=False)
```

Comment:
make sure all instances of `save_granular_log` are wrapped in the same `with granular_lock:`

Reply:
I see.  It might be better to add the lock in the sota.py save_granular_log function instead of the function in UpdateLogger class.

Reply:
> I see. It might be better to add the lock in the sota.py save_granular_log function instead of the function in UpdateLogger class as the dispatcher will be calling the function there.

the problem with doing it there is that you still need to lock the 'read-update-write' transaction

Reply:
Oh I see. It is possible that other threads may access the UpdateLogger class. I will also add another lock in the UpdateLogger class